### ISSUE TITLE
Clarify if `ProgressEvent.loaded` should indicate the size of compress, or uncompressed, data

### ISSUE THREAD

**Issue by Zwyx** [whatwg/xhr#388]
### What is the issue with the XMLHttpRequest Standard?

When a [`ProgressEvent`](https://xhr.spec.whatwg.org/#interface-progressevent) fires, `ProgressEvent.loaded` contains the size of the data already transmitted.

If the data is compressed and its total size is unknown, then, **depending on the browser, `ProgressEvent.loaded` can contain either the size of the compressed data, or the size of the decompressed data**.

It would be great if the spec could clarify which of these two values should be the one provided by `ProgressEvent.loaded`.

### Additional context

With a response containing compressed data and unknown total size (the `Content-Length` header is not set), `ProgressEvent.lengthComputable` is `false` and `ProgressEvent.total` is not usable, preventing us from displaying a progress indicator.

In this case, a [common technique](https://stackoverflow.com/questions/15097712/how-can-i-use-deflated-gzipped-content-with-an-xhr-onprogress-function/32799706#32799706) is to set a custom response header `X-Uncompressed-Content-Length` containing the size of the uncompressed data being transmitted.

We can then, on the client side, use the value of this header as a substitue of `ProgressEvent.total`, and display a progress indicator with the value of `ProgressEvent.loaded` divided by the value of this header.

However, depending on the browser, `ProgressEvent.loaded` can be either the size of the compressed data already transmitted (the case in Firefox), or, the size of the decompressed data already transmitted (the case in Chrome).

For example:
- In Chrome, downloading a content of 1.3MB (1,345,276 bytes), while logging the value of `loaded`, shows a total of 1345276 at the end.
- In Firefox, downloading the same content of 1,345,276 bytes, while logging the value of `loaded`, shows a total of 1017454 at the end.

**Comment by annevk** [comment]
Firefox is correct. This is actually defined in detail through the interaction with Fetch. Content codings are removed before the creation of a response stream.

**Comment by Zwyx** [comment]
Thank you for replying. Would you be able to provide a link to where this is defined? I searched the [Fetch spec](https://fetch.spec.whatwg.org/#concept-fetch) for `progress`, `leaded`, `compress`, etc. but didn't find anything.

I'm also surprised the Fetch API has anything to do in XMLHttpRequest :thinking:, considering Fetch came around more than a decade later?

**Comment by cutiful** [comment]
@annevk is it still true that Firefox is correct? It seems like https://github.com/whatwg/xhr/pull/313 would change that. As far as I understand, transmitted bytes on body referenced the compressed length here:
```html
 <li><p>Let <var>transmitted</var> be <var>response</var>'s <a for=response>body</a>'s
 <a for=body>transmitted bytes</a>.
```

But in the new version, it uses the length of received bytes:
```html
 <li><p>Let <var>transmitted</var> be <var>xhr</var>'s <a>received bytes</a>'s
 <a for="byte sequence">length</a>.
```

Which, if I'm not mistaken, are set by the HTTP-network fetch algorithm after handling codings. Otherwise, the XHR specification would have to handle codings as well, so I believe my understanding is correct that this is decompressed length.

That makes the Chrome behavior the correct one. And I think it also means that the progress event is potentially fired with mismatching lengths (decompressed loaded and compressed total).

**Comment by annevk** [comment]
Yeah, I'm not sure why I suggested above that Firefox would be correct, as I said at the same time that the stream is created post-content-codings, which would mean decompressed.

**Comment by cutiful** [comment]
Thanks for the quick response! In that case, is setting length/total to Content-Length for compressed responses a bug? Should an issue be created for it?

**Comment by annevk** [comment]
I don't think there's anything else we can do? We could update it once we have the full response perhaps, but that might also break someone's logic at this point. Probably best discussed separately, yes.
